File: android/app/src/main/kotlin/com/elnemr/language/CacheCleaner.kt
package com.elnemr.language

import android.content.Context
import io.flutter.plugin.common.MethodChannel
import java.io.File

/// Clears the app's on-disk cache (`context.cacheDir`) on demand — the UTF-8
/// subtitle re-encode temp files (`elnemr_sub_*.utf8`) and anything else
/// the engine wrote there. Android may purge this directory at any time, so it
/// is always safe to delete; ExoPlayer keeps no disk cache of its own. The
/// in-memory image cache (TMDB posters/backdrops/stills) lives in Flutter and
/// is cleared from Dart (`CacheCleaner.clearMemoryImages`).
class CacheCleaner(private val context: Context) {

    fun configure(channel: MethodChannel) {
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "size" -> result.success(cacheSizeBytes())
                "clear" -> result.success(clearCacheBytes())
                else -> result.notImplemented()
            }
        }
    }

    private fun cacheDir(): File = context.cacheDir

    private fun cacheSizeBytes(): Long =
        cacheDir().walkBottomUp().filter { it.isFile }.sumOf { it.length() }

    private fun clearCacheBytes(): Long {
        var freed = 0L
        for (file in cacheDir().walkBottomUp()) {
            if (file.isFile) {
                freed += file.length()
                file.delete()
            } else if (file != cacheDir()) {
                file.delete()
            }
        }
        return freed
    }
}

File: ios/Runner/CacheCleaner.swift
import Flutter
import Foundation

/// Clears the app's on-disk cache on demand (the `elnemr/cache` channel).
/// Deletes the contents of the app's Caches directory and tmp — subtitle temp
/// files and anything AetherEngine/FFmpeg dropped there. Both directories are
/// OS-purgeable, so deleting them is always safe. The in-memory image cache
/// (TMDB posters/backdrops/stills) lives in Flutter and is cleared from Dart
/// (`CacheCleaner.clearMemoryImages`).
enum CacheCleaner {

    static func register(with messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(
            name: "elnemr/cache",
            binaryMessenger: messenger
        )
        channel.setMethodCallHandler { call, result in
            switch call.method {
            case "size":
                result(diskSizeBytes())
            case "clear":
                result(clearCacheBytes())
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    private static var cacheDirectories: [URL] {
        var urls: [URL] = []
        if let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first {
            urls.append(caches)
        }
        urls.append(FileManager.default.temporaryDirectory)
        return urls
    }

    private static func diskSizeBytes() -> Int64 {
        var total: Int64 = 0
        for dir in cacheDirectories {
            if let enumerator = FileManager.default.enumerator(at: dir, includingPropertiesForKeys: [.fileSizeKey, .isDirectoryKey]) {
                for case let url as URL in enumerator {
                    guard let values = try? url.resourceValues(forKeys: [.fileSizeKey, .isDirectoryKey]),
                          values.isDirectory != true,
                          let size = values.fileSize else { continue }
                    total += Int64(size)
                }
            }
        }
        return total
    }

    private static func clearCacheBytes() -> Int64 {
        var freed: Int64 = 0
        for dir in cacheDirectories {
            guard let enumerator = FileManager.default.enumerator(
                at: dir,
                includingPropertiesForKeys: [.isDirectoryKey],
                options: [.skipsHiddenFiles]
            ) else { continue }
            for case let url as URL in enumerator {
                let isDirectory = (try? url.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true
                if isDirectory {
                    continue
                }
                let size = (try? url.resourceValues(forKeys: [.fileSizeKey]))?.fileSize ?? 0
                if (try? FileManager.default.removeItem(at: url)) != nil {
                    freed += Int64(size)
                }
            }
        }
        return freed
    }
}
